src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -989,98 +938,178 @@ HeapWord* G1CollectedHeap::satisfy_failed_allocation_helper(size_t word_size,
+  }
+
+  // Attempt to satisfy allocation requests after full gc also failed. We reset the allocation requests
+  // then execute a maximal compaction full gc before retrying the allocations.
+  reset_allocation_requests();
+
+  *gc_succeeded = do_full_collection(true /* clear_all_soft_refs */, true /* do_maximal_compaction */);
+
+  if (!*gc_succeeded) {
+    return false;
+  }
+
+  alloc_succeeded = handle_allocation_requests(true /* expect_null_mutator_alloc_region*/);
+
+  if (alloc_succeeded) {
+    return alloc_succeeded;
+  }
+
+  // Even after maximal compaction full gc, we failed to satisfy all the pending allocation requests.
+  // We can declare the pending requests as failed.
+  DoublyLinkedList<StalledAllocReq>::RemoveIterator iter(&_stalled_allocations);
+
+  for (StalledAllocReq* alloc_req; iter.next(&alloc_req);) {
+    alloc_req->set_state(StalledAllocReq::AllocationState::Failed);
+    _satisfied_allocations.insert_last(alloc_req);
   }
 
   assert(!soft_ref_policy()->should_clear_all_soft_refs(),
          "Flag should have been handled and cleared prior to this point");
+  return false;
+}
 
-  // What else?  We might try synchronous finalization later.  If the total
-  // space available is large enough for the allocation, then a more
-  // complete compaction phase than we've tried so far might be
-  // appropriate.
-  return nullptr;
+void G1CollectedHeap::reset_allocation_requests() {
+  assert_at_safepoint_on_vm_thread();
+
+  DoublyLinkedList<StalledAllocReq> prev_allocations;
+
+  prev_allocations.swap(_satisfied_allocations);
+
+  assert(_satisfied_allocations.is_empty(), "Sanity check!");
+
+  DoublyLinkedList<StalledAllocReq>::RemoveIterator iter(&prev_allocations);
+
+  for (StalledAllocReq* alloc_req; iter.next(&alloc_req);) {
+    if (alloc_req->state() == StalledAllocReq::AllocationState::Failed) {
+      // If an allocation request was declared failed, we maintain the failed state.
+      // This prevents requests from flip-floping between Pending->Failed->Pending states.
+      // Once any allocation attempt declares a request as failed, that request will not be re-attempted.
+      _satisfied_allocations.insert_first(alloc_req);
+    } else {
+      alloc_req->set_state(StalledAllocReq::AllocationState::Pending);
+      _stalled_allocations.insert_first(alloc_req);
+    }
+  }
 }
 
-// Attempting to expand the heap sufficiently
-// to support an allocation of the given "word_size".  If
-// successful, perform the allocation and return the address of the
-// allocated block, or else null.
+bool G1CollectedHeap::handle_allocation_requests(bool expect_null_mutator_alloc_region) {
+  assert_at_safepoint_on_vm_thread();
+
+  const uint active_numa_nodes = G1NUMA::numa()->num_active_nodes();
+  bool *expect_null_alloc_regions = (bool*)alloca(active_numa_nodes * sizeof(bool));
+  for (uint i = 0; i < active_numa_nodes; i++) {
+    expect_null_alloc_regions[i] = expect_null_mutator_alloc_region;
```

Comment:
I wasn't in position to determine that, so I went with keeping it. Removing it would make the code a lot cleaner and also avoid the use of alloca, but again, I need guidance on the removing it part